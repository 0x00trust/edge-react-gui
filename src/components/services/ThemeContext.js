// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig.js'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
