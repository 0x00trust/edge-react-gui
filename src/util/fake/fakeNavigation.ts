import { NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popTo: () => {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  getState() {
    return {
      drawerOpen: false,
      index: 0,
      routes: []
    }
  }
}
