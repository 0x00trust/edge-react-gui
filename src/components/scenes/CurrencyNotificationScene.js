// @flow

import * as React from 'react'
import { ScrollView } from 'react-native'
import { sprintf } from 'sprintf-js'

import { newPriceChangeEvent, serverSettingsToState, setDeviceSettings } from '../../actions/NotificationActions.js'
import { type NewPushEvent } from '../../controllers/action-queue/types/pushTypes.js'
import { useHandler } from '../../hooks/useHandler.js'
import s from '../../locales/strings.js'
import { useCallback, useMemo } from '../../types/reactHooks.js'
import { useDispatch, useSelector } from '../../types/reactRedux.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { SceneWrapper } from '../common/SceneWrapper.js'
import { showError } from '../services/AirshipInstance'
import { SettingsSwitchRow } from '../themed/SettingsSwitchRow.js'

type OwnProps = {
  navigation: NavigationProp<'currencyNotificationSettings'>,
  route: RouteProp<'currencyNotificationSettings'>
}

type Props = OwnProps

export const CurrencyNotificationScene = (props: Props) => {
  const { route } = props
  const { currencyInfo } = route.params
  const { pluginId } = currencyInfo
  const dispatch = useDispatch()

  const defaultIsoFiat = useSelector(state => state.ui.settings.defaultIsoFiat)
  const settings = useSelector(state => state.priceChangeNotifications)

  const toggleHourlySetting = useHandler(async () => {
    const newEvent = newPriceChangeEvent(currencyInfo, defaultIsoFiat, !settings[pluginId].hourlyChange, !!settings[pluginId].dailyChange)
    await updateSettings(newEvent)
  })

  const toggleDailySetting = useHandler(async () => {
    const newEvent = newPriceChangeEvent(currencyInfo, defaultIsoFiat, !!settings[pluginId].hourlyChange, !settings[pluginId].dailyChange)
    await updateSettings(newEvent)
  })

  const updateSettings = useCallback(
    async (event: NewPushEvent) => {
      try {
        const newSettings = await dispatch(setDeviceSettings({ createEvents: [event] }))
        dispatch({
          type: 'PRICE_CHANGE_NOTIFICATIONS_UPDATE',
          data: serverSettingsToState(newSettings)
        })
      } catch (e) {
        showError(`Failed to reach notification server: ${e}`)
      }
    },
    [dispatch]
  )

  const rows = useMemo(
    () => [
      <SettingsSwitchRow
        key="hourly"
        label={sprintf(s.strings.settings_currency_notifications_percent_change_hour, 3)}
        value={settings[pluginId].hourlyChange != null}
        onPress={toggleHourlySetting}
      />,
      <SettingsSwitchRow
        key="daily"
        label={sprintf(s.strings.settings_currency_notifications_percent_change_hours, 10, 24)}
        value={settings[pluginId].dailyChange != null}
        onPress={toggleDailySetting}
      />
    ],
    [pluginId, settings, toggleDailySetting, toggleHourlySetting]
  )

  return (
    <SceneWrapper background="theme" hasTabs={false}>
      <ScrollView>{rows}</ScrollView>
    </SceneWrapper>
  )
}
