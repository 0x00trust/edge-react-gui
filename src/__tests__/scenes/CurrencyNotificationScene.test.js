/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { CurrencyNotificationComponent } from '../../components/scenes/CurrencyNotificationScene.js'
import { getTheme } from '../../components/services/ThemeContext.js'
import { fakeNavigation } from '../../util/fake/fakeNavigation.js'

describe('CurrencyNotificationComponent', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      navigation: fakeNavigation,
      route: {
        params: {
          currencyInfo: {
            pluginId: '',
            displayName: 'Bitcoin',
            walletType: 'My Bitcoin Wallet',
            currencyCode: 'BTC',
            denominations: [
              {
                multiplier: '10000000',
                name: 'Sats'
              }
            ],
            addressExplorer: 'Blockchair',
            transactionExplorer: 'Blockchair',
            defaultSettings: ['JsonSettings'],
            metaTokens: [
              {
                currencyCode: 'BTC',
                currencyName: 'Bitcoin',
                denominations: [
                  {
                    multiplier: '10000000',
                    name: 'Sats'
                  }
                ]
              }
            ]
          }
        }
      },

      userId: '',
      disklet: {
        state: {
          core: {
            disklet: () => undefined
          }
        }
      },
      enableNotifications: (currencyCode, hours, enabled) => undefined,
      theme: getTheme()
    }

    const actual = renderer.render(<CurrencyNotificationComponent {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
