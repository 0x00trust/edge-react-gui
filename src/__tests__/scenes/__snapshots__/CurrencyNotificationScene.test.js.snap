// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyNotificationComponent should render with loading props 1`] = `
<SceneWrapper
  background="theme"
  hasTabs={false}
>
  <ScrollView />
</SceneWrapper>
`;
